File: ios/LeaderDojo/LeaderDojo/Core/Services/EntriesService.swift
import Foundation

struct SummarizeEntryResponse: Decodable {
    let summary: String
    let suggestedActions: [Entry.SuggestedAction]
}

struct EntrySuggestionInput: Encodable {
    let projectId: String
    let actions: [Action]

    struct Action: Encodable {
        let title: String
        let direction: Commitment.Direction
        let counterparty: String?
        let dueDate: Date?
        let importance: Int?
        let urgency: Int?
        let notes: String?
    }
}

struct EntriesService {
    let apiClient: APIClient

    init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    func listEntries(projectId: String) async throws -> [Entry] {
        try await apiClient.request(.projectEntries(id: projectId), responseType: [Entry].self)
    }

    func createEntry(projectId: String, input: CreateEntryInput) async throws -> Entry {
        try await apiClient.request(.createEntry(projectId: projectId, payload: AnyEncodable(input)), responseType: Entry.self)
    }

    func summarizeEntry(entryId: String) async throws -> SummarizeEntryResponse {
        try await apiClient.request(.summarizeEntry(entryId: entryId), responseType: SummarizeEntryResponse.self)
    }

    func createCommitmentsFromSuggestions(entryId: String, payload: EntrySuggestionInput) async throws -> [Commitment] {
        try await apiClient.request(.createCommitments(entryId: entryId, payload: AnyEncodable(payload)), responseType: [Commitment].self)
    }
}

File: web/src/lib/services/entries-service.ts
import type { SuggestedAction } from "@/lib/ai/types";
import { getAiClient } from "@/lib/ai/client";
import type { NewProjectEntry, ProjectEntry } from "@/lib/db/types";
import { CommitmentsRepository, EntriesRepository, ProjectsRepository } from "@/lib/repositories";

type CreateEntryInput = Omit<
  NewProjectEntry,
  "id" | "userId" | "projectId" | "createdAt" | "updatedAt"
>;

export class EntriesService {
  constructor(
    private readonly entriesRepo = new EntriesRepository(),
    private readonly projectsRepo = new ProjectsRepository(),
    private readonly commitmentsRepo = new CommitmentsRepository(),
  ) {}

  async createEntry(userId: string, projectId: string, input: CreateEntryInput) {
    const entry = await this.entriesRepo.createEntry({
      ...input,
      userId,
      projectId,
    });

    await this.projectsRepo.updateLastActive(userId, projectId, input.occurredAt ?? new Date());
    return entry;
  }

  async getTimeline(userId: string, projectId: string) {
    return this.entriesRepo.listByProject(userId, projectId);
  }

  async markAsDecision(userId: string, entryId: string) {
    return this.entriesRepo.updateEntry(userId, entryId, { isDecision: true });
  }

  async getEntry(userId: string, entryId: string) {
    return this.entriesRepo.findById(userId, entryId);
  }

  async updateEntry(
    userId: string,
    entryId: string,
    input: Partial<CreateEntryInput>
  ) {
    return this.entriesRepo.updateEntry(userId, entryId, input);
  }

  async deleteEntry(userId: string, entryId: string) {
    return this.entriesRepo.deleteEntry(userId, entryId);
  }

  async generateSummary(entry: ProjectEntry, projectContext?: string) {
    const ai = getAiClient();
    return ai.summarizeMeeting({
      rawContent: entry.rawContent ?? "",
      projectContext,
    });
  }

  async persistSummary(
    userId: string,
    entryId: string,
    summary: string,
    suggestedActions: SuggestedAction[],
  ) {
    await this.entriesRepo.updateEntry(userId, entryId, {
      aiSummary: summary,
      aiSuggestedActions: suggestedActions,
    });
  }

  async createCommitmentsFromSuggestions(
    userId: string,
    projectId: string,
    entryId: string,
    suggestions: SuggestedAction[],
  ) {
    const payload = suggestions.map((suggestion) => ({
      userId,
      projectId,
      entryId,
      title: suggestion.title,
      direction: suggestion.direction,
      counterparty: suggestion.counterparty,
      dueDate: suggestion.dueDate ? new Date(suggestion.dueDate) : null,
      importance: suggestion.importance ?? 3,
      urgency: suggestion.urgency ?? 3,
      notes: suggestion.notes,
      aiGenerated: true,
    }));

    return this.commitmentsRepo.createMany(payload);
  }

  /**
   * Get all entries for a user with their project information
   * Used for the cross-project activity timeline
   */
  async getAllEntriesWithProjects(
    userId: string,
    filters?: {
      projectId?: string;
      kinds?: ProjectEntry["kind"][];
      limit?: number;
    }
  ) {
    const entries = await this.entriesRepo.listForUser(userId, {
      projectId: filters?.projectId,
      kinds: filters?.kinds,
    });

    // Get unique project IDs
    const projectIds = [...new Set(entries.map((e) => e.projectId))];
    
    // Fetch all projects in parallel
    const projectsMap = new Map<string, { name: string; id: string }>();
    await Promise.all(
      projectIds.map(async (projectId) => {
        const project = await this.projectsRepo.findById(userId, projectId);
        if (project) {
          projectsMap.set(projectId, { name: project.name, id: project.id });
        }
      })
    );

    // Combine entries with project info
    const entriesWithProjects = entries
      .map((entry) => ({
        ...entry,
        projectName: projectsMap.get(entry.projectId)?.name ?? "Unknown Project",
      }))
      .slice(0, filters?.limit ?? 50);

    return entriesWithProjects;
  }
}

